src/coreclr/jit/block.h

```diff
@@ -552,7 +552,8 @@ enum BasicBlockFlags : unsigned __int64
     BBF_PATCHPOINT           = MAKE_BBFLAG(36), // Block is a patchpoint
     BBF_HAS_CLASS_PROFILE    = MAKE_BBFLAG(37), // BB contains a call needing a class profile
     BBF_PARTIAL_COMPILATION_PATCHPOINT  = MAKE_BBFLAG(38), // Block is a partial compilation patchpoint
-    BBF_HAS_ALIGN          = MAKE_BBFLAG(39), // BB ends with 'align' instruction
+    BBF_HAS_ALIGN            = MAKE_BBFLAG(39), // BB ends with 'align' instruction
+    BBF_TAILCALL_SUCCESSOR   = MAKE_BBFLAG(40), // BB has pred that has potential tail call
```

Comment:
I assume you meant `BB has successor` ?

Reply:
No -- it marks blocks that come after tail calls. 

Maybe a picture will help? Here's a fragment of an OSR method flow graph before we add instrumentation. We want to count how often `R` is executed, but we can't put probes in `R` because it is marked with `BBF_TAILCALL_SUCCESSOR` -- it needs to remain empty since the tail call preds won't execute `R`.

Also pictured are some non-tail call blocks `A` and `B` that conditionally share the return, and an OSR-unreachable block `Z`. And the blue edge is a fall-through edge. `A` has degenerate flow, which is rare, but possible.

![image - 2021-12-02T085013 213](https://user-images.githubusercontent.com/10121823/144466524-62a1b5ef-af68-43a9-96cc-4f861854effe.png)

To handle this we need to put copies of `R`'s probes in the tail call blocks, and create an intermediary block that all the other preds flow through to get to `R`. So we end up with 3 separate copies of R's pgo probe that collectively give us the right count for `R`, and `R` remains empty so the tail calls work as expected.

![image - 2021-12-02T085126 253](https://user-images.githubusercontent.com/10121823/144466710-0e1aa59e-e86d-4064-9a09-06d48a45e0fb.png)

We also take pains not to instrument `Z`, since there are debug checks that verify that un-imported blocks remain empty and can be removed. And we take pains not to double-count `A`.

Reply:
Oh, yeah it makes sense now for me, thanks for detailed response 🙂 not going to mark it as resolved to keep it.